gdal/ogr/ogrsf_frmts/ngw/ogrngwlayer.cpp

```diff
@@ -0,0 +1,1116 @@
+ */
+std::string OGRNGWLayer::GetResourceId() const
+{
+    return osResourceId;
+}
+
+/*
+ * Delete()
+ */
+bool OGRNGWLayer::Delete()
+{
+    if( osResourceId == "-1")
+    {
+        return true;
+    }
+
+    // Headers free in DeleteResource method.
+    return NGWAPI::DeleteResource(poDS->GetUrl(), osResourceId, poDS->GetHeaders());
+}
+
+/*
+ * ResetReading()
+ */
+void OGRNGWLayer::ResetReading()
+{
+    SyncToDisk();
+    if( poDS->GetPageSize() != -1 )
+    {
+        FreeFeaturesCache();
+        nPageStart = 0;
+    }
+    oNextPos = moFeatures.begin();
+}
+
+/*
+ * SetNextByIndex()
+ */
+OGRErr OGRNGWLayer::SetNextByIndex( GIntBig nIndex )
+{
+    SyncToDisk();
+    if( nIndex < 0)
+    {
+        CPLError(CE_Failure, CPLE_AppDefined,
+            "Feature index must be greater or equal 0. Got " CPL_FRMT_GIB, nIndex);
+        return OGRERR_FAILURE;
+    }
+    if( poDS->GetPageSize() != -1 )
+    {
+        if( nPageStart < nIndex && nIndex >= nPageStart - poDS->GetPageSize() )
+        {
+            oNextPos = moFeatures.begin();
+            std::advance(oNextPos, nIndex);
+        }
+        else
+        {
+            nPageStart = nIndex / poDS->GetPageSize() * poDS->GetPageSize();
+            oNextPos = moFeatures.end();
+        }
+    }
+    else
+    {
+        oNextPos = moFeatures.begin();
+        std::advance(oNextPos, nIndex);
+    }
+    return OGRERR_NONE;
+}
+
+/*
+ * GetNextFeature()
+ */
+OGRFeature *OGRNGWLayer::GetNextFeature()
+{
+    std::string osUrl;
+    if( poDS->GetPageSize() != -1 )
+    {
+        if( oNextPos == moFeatures.end() && nPageStart < GetFeatureCount(FALSE) )
+        {
+            FreeFeaturesCache();
+
+            osUrl = NGWAPI::GetFeaturePage( poDS->GetUrl(), osResourceId,
+                nPageStart, poDS->GetPageSize() );
+            nPageStart += poDS->GetPageSize();
+        }
+    }
+    else if( moFeatures.empty() && GetFeatureCount(FALSE) > 0 )
+    {
+        osUrl = NGWAPI::GetFeature( poDS->GetUrl(), osResourceId );
+    }
+
+    if( !osUrl.empty() )
+    {
+        CPLErrorReset();
+        CPLJSONDocument oFeatureReq;
+        char **papszHTTPOptions = poDS->GetHeaders();
+        bool bResult = oFeatureReq.LoadUrl( osUrl, papszHTTPOptions );
+        CSLDestroy( papszHTTPOptions );
+
+        CPLJSONObject oRoot = oFeatureReq.GetRoot();
+        if( !CheckRequestResult(bResult, oRoot, "GetFeatures request failed") )
+        {
+            return nullptr;
+        }
+
+        CPLJSONArray aoJSONFeatures = oRoot.ToArray();
+        for( int i = 0; i < aoJSONFeatures.Size(); ++i)
+        {
+            OGRFeature *poFeature = JSONToFeature( aoJSONFeatures[i],
+                poFeatureDefn, false );
+            moFeatures[poFeature->GetFID()] = poFeature;
+        }
+
+        oNextPos = moFeatures.begin();
+
+        // Set m_nFeaturesRead for GetFeaturesRead.
+        if( poDS->GetPageSize() != -1 )
+        {
+            m_nFeaturesRead = moFeatures.size();
```

Comment:
shouldn't that be += ?

Reply:
In this place paging is off and we receive all features at once. So  moFeatures size is total features read.